net.solarnetwork.node.datum.energymeter.mock/src/net/solarnetwork/node/datum/energymeter/mock/MockEnergyMeterDatumSource.java

```diff
@@ -0,0 +1,287 @@
+		calcRandomness();
+
+		// the values for most datum variables are calculated here
+		calcVariables();
+
+		// calculates the watt hours
+		calcWattHours();
+
+		GeneralNodeACEnergyDatum datum = new GeneralNodeACEnergyDatum();
+
+		datum.setCreated(new Date());
+		datum.setSourceId(sourceId);
+		datum.setFrequency((float) this.randomFreq);
+		datum.setVoltage((float) this.randomVolt);
+		datum.setReactivePower(this.reacPow);
+		datum.setApparentPower(this.apparantPow);
+		datum.setPhaseVoltage(this.phaseVolt);
+		datum.setCurrent(this.phaseCurr);
+		datum.setWatts(this.watt);
+		datum.setRealPower(this.realPow);
+		datum.setPowerFactor(this.powfac);
+		datum.setWattHourReading(this.watthours);
+		return datum;
+	}
+
+	private void calcRandomness() {
+		this.randomVolt = Double.parseDouble(voltagerms);
+		this.randomFreq = Double.parseDouble(frequency);
+
+		// if randomness is off randomVolt and randomFreq will have no deviation
+		if (Boolean.parseBoolean(randomness)) {
+
+			// add deviation to the supply
+			double vd = Double.parseDouble(this.voltDeviation);
+			double fd = Double.parseDouble(this.freqDeviation);
+			this.randomVolt += vd * Math.cos(Math.PI * rng.nextDouble());
+			this.randomFreq += fd * Math.cos(Math.PI * rng.nextDouble());
+		}
+	}
+
+	/**
+	 * Calculates the values to feed the datum
+	 */
+	private void calcVariables() {
+		double vrms = this.randomVolt;
+
+		// convention to use capital L for inductance reading in microhenry
+		double L = Double.parseDouble(inductance) / 1000000;
+
+		double f = this.randomFreq;
+
+		// convention to use capital R for resistance
+		double R = Double.parseDouble(resistance);
+
+		double vmax = Math.sqrt(2) * vrms;
+
+		double phasevoltage = vmax * Math.sin(2 * Math.PI * f * System.currentTimeMillis() / 1000);
+		this.phaseVolt = (float) phasevoltage;
+
+		double inductiveReactance = 2 * Math.PI * f * L;
+		double impedance = Math.sqrt(Math.pow(R, 2) + Math.pow(inductiveReactance, 2));
+
+		double phasecurrent = phasevoltage / impedance;
+		this.phaseCurr = (float) phasecurrent;
+		double current = vrms / impedance;
+
+		double reactivePower = Math.pow(current, 2) * inductiveReactance;
+		this.reacPow = (int) reactivePower;
+		double realPower = Math.pow(current, 2) * R;
+		this.realPow = (int) realPower;
+		this.apparantPow = (int) (Math.pow(current, 2) * impedance);
+
+		// not sure if correct calculation
+		double watts = Math.pow(phasecurrent, 2) * R;
+		this.watt = (int) watts;
+
+		double phaseAngle = Math.atan(inductiveReactance / R);
+		this.powfac = (float) Math.cos(phaseAngle);
+
+	}
+
+	private void calcWattHours() {
+		if (this.lastsample == null) {
+			this.lastsample = System.currentTimeMillis();
+			this.wattmillis = 0L;
+			this.watthours = 0L;
+		} else {
+			Long newsample = System.currentTimeMillis();
+			Long diff = (newsample - this.lastsample);
+			this.wattmillis += this.reacPow.longValue() * diff;
```

Comment:
So for watt hour calculation, I had expected to see that calculated from real power, not reactive power. I'm no expert so might want to check that with Philip or Ed.

Reply:
I also intended to use real power thats a typo thank you for spotting.